CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1643 @@
+directly accesses the module in the `i`th component of the global
+module environment. This reduces the cost from a variable lookup plus
+a projection, to just a projection; this can be expected to speed up every
+reference to an external module.
+
+##### Subvariation: Unboxed modules
+
+This subvariation makes every reference to a module export cheaper, by
+replacing two projections from a tuple by one. It does require
+preprocessing script code before it is run, updating export references
+to refer to the correct element of the large tuple combining several
+modules. This requires a traversal of all the script code in a
+transaction, which must be performed every time script verification is
+run, including on the chain. Because of this, it makes most sense to
+use this subvariation in combination with 'global module environment',
+which also requires such a traversal. In both cases, the purpose is to
+adjust references to refer to the correct index in the new, merged
+data structure; a single traversal suffices to achieve both ends.
+
+The syntactic restriction, requiring a module body to be a tuple of
+exports, is not onerous. While some compilers might wish to represent
+a module as built-in data, or as a function from a tag (as Solidity
+does), this can be achieved by placing the intended module value as
+the first component of a one-tuple. The implementation described here
+optimises away the selection of an element from such a tuple, so the
+restriction introduces no extra overhead in this case.
+
+### Transaction fees
+
+Imported modules are provided using reference scripts, an existing
+mechanism (see CIP-33), or in the transaction itself. Provided the
+cost of loading reference scripts is correctly accounted for, this CIP
+introduces no new problems.
+
+Note that there is now (since August 2024) a hard limit on the total
+size of reference scripts used in a transaction, and the transaction
+fee is exponential in the total script size (see
+[here](https://github.com/IntersectMBO/cardano-ledger/blob/master/docs/adr/2024-08-14_009-refscripts-fee-change.md)).The
+exponential fees provide a strong motivation to prefer the 'lazy
+loading' variation in this CIP: even a small reduction in the number
+of reference scripts that need to be provided may lead to a large
+reduction in transaction fees.
+
+The motivation for these fees is to deter DDoS attacks based on
+supplying very large Plutus scripts that are costly to deserialize,
+but run fast and so incur low execution unit fees. While these fees
+are likely to be reasonable for moderate use of the module system, in
+the longer term they could become prohibitive for more complex
+applications. It may be necessary to revisit this design decision in
+the future. To be successful, the DDoS defence just needs fees to
+become *sufficiently* expensive per byte as the total size of
+reference scripts grows; they do not need to grow without bound. So
+there is scope for rethinking here.

```

Comment:
> It may be necessary to revisit this design decision in the future.

I don't think this can be left for "future work". I really think it should be updated if necessary when this CIP gets implemented. The reason for this is I don't think DApps should be treated as standalone applications. I think the following example perfectly exemplifies why:

Right now, all stablecoins are not fungible despite them all effectively being the the US dollar. You can't repay a loan in DJED using USDM. If DApps were composable, you could compose a DEX with the lending/borrowing DApp to convert the USDM to DJED in the same transaction where you make the loan payment. DApp composability makes stablecoins fungible!

This isn't possible on account style blockchains because each DApp is individually too expensive. On Cardano, you can compose 10 different DApps in the same transaction. I think this module approach would be huge, but only if it doesn't interfere with DApp composability. AFAIU that means lazy loading is 100% a requirement and users should be able to compose 4-5 DApps in a single transaction even with this module approach. Otherwise, this CIP could end up seriously handicapping the potential of Cardano's DeFi. 

I was personally frustrated when I saw there was a hard-cap on the reference script size; if people want to pay up to fit more DApps into the transaction, let them! I'm fine with the cost being exponential after a certain point (ideally after 4-5 DApps in the transaction), but the hard limit doesn't make sense to me as long as the user pays for it. The adr linked to doesn't give any justification for the hard limit aside from "further increase the resilience". This CIP could easily exacerbate the issues with the reference script fee calculation.

Reply:
I agree it's going to be necessary. I just don't think it's a *prerequisite*... so modules should not be held up waiting for this. They'll be useful even without a change to reference script fees--just not as useful. I realise there are other factors to consider in fee-setting, but adding modules should raise the priority of fixing those fees considerably.